In [6]:
import nltk

%autoreload 2

UsageError: Line magic function `%autoreload` not found.


In [104]:
from ass6utils import read_file, score, MeaningRepresentation, bleu_single
import random

meaning_representations, references = read_file('devset.csv')

def generate(mr: MeaningRepresentation) -> str:
    return '{} is a {} {}.'.format(
        mr.name,
        mr.food,
        mr.eat_type,
    )

def no_none(word, prefix='', replace=None, postfix='', default = None):
    if default:
        word = default
    if word == None:
        return ''
    if replace:
        if word == 'no':
            return prefix + 'non ' + replace + postfix + ' '
        else:
            return prefix + replace + postfix + ' '
    return prefix + word + postfix + ' '

def generate(mr: MeaningRepresentation) -> str:
    return f'{no_none(mr.name)}is a {no_none(mr.family_friendly, replace="family-friendly")}{no_none(mr.customer_rating, postfix=" rated")}{no_none(mr.price_range, postfix=" priced")} {no_none(mr.food)}restaurant {no_none(mr.eat_type)}{no_none(mr.near, prefix=" near ")}{no_none(mr.area, prefix="in ")}.'

score(generate, meaning_representations, references)
print('--')
for _ in range(10):
    print('\t', generate(random.choice(meaning_representations)))
print('\n')


BLEU score: 0.19901693177241017
ROUGE-L score: 0.5755707158816794
--
	 The Cricketers is a family-friendly low rated  Chinese restaurant coffee shop  near The Portland Arms .
	 Fitzbillies is a non family-friendly 1 out of 5 rated high priced  Chinese restaurant coffee shop in city centre .
	 The Golden Palace is a 1 out of 5 rated high priced  Chinese restaurant coffee shop in city centre .
	 The Punter is a family-friendly 5 out of 5 rated  restaurant in city centre .
	 Clowns is a low rated  English restaurant coffee shop  near Clare Hall in city centre .
	 Fitzbillies is a non family-friendly 1 out of 5 rated moderate priced  Chinese restaurant coffee shop in riverside .
	 The Mill is a moderate priced  English restaurant coffee shop  near The Sorrento in city centre .
	 The Wrestlers is a family-friendly 3 out of 5 rated  restaurant  near The Sorrento .
	 Browns Cambridge is a non family-friendly low rated  Chinese restaurant coffee shop  near Crowne Plaza Hotel in riverside .
	 F

1. On a basic level the task is not too difficult if the goal is to just generate that x is a z type thing in y.
2. Well the scores by themselves don't really mean anything but when they are combined with a task such as this and to other systems they have the ability to somewhat rank the systems but they don't really explain exactly how good a human would think the system is.
3. The devset represents very formulaic type of text that could be easily represented as rules.

1. The meteor score did not work for me, so I don't get the meteor scores. The default generates BLEU score of 0.03 and rouge score of 0.38.
2. These are very poor results compared to the baseline.
3. Main problem is the None outputs when the word does not exist.

Adding some rules I got the following scores and sentences.

BLEU score: 0.19901693177241017
ROUGE-L score: 0.5755707158816794


	 The Wrestlers is a non family-friendly 5 out of 5 rated  restaurant  near The Sorrenato .
	 Cocum is a non family-friendly 1 out of 5 rated moderate priced  English restaurant coffee shop .
	 The Eagle is a non family-friendly high rated £20-25 priced  English restaurant coffee shop  near Burger King in riverside .
	 Fitzbillies is a non family-friendly 5 out of 5 rated cheap priced  Chinese restaurant coffee shop in riverside .
	 The Wrestlers is a non family-friendly less than £20 priced  English restaurant coffee shop  near Raja Indian Cuisine in riverside .
	 Fitzbillies is a non family-friendly average rated cheap priced  Chinese restaurant coffee shop in riverside .
	 Clowns is a high rated  English restaurant coffee shop  near Clare Hall in riverside .
	 The Eagle is a family-friendly low rated less than £20 priced  Chinese restaurant coffee shop  near Burger King in riverside .
	 Cotto is a 3 out of 5 rated moderate priced  Chinese restaurant coffee shop  near The Portland Arms in city centre .
	 Fitzbillies is a family-friendly 1 out of 5 rated high priced  English restaurant coffee shop in city centre .


In [81]:
def no_none(word, prefix='', replace=None, postfix=''):
    if word == None:
        return ''
    if replace:
        return prefix + replace + postfix + ' '
    return prefix + word + postfix + ' '

def generate(mr: MeaningRepresentation) -> str:
    return f'{no_none(mr.name)}is a {no_none(mr.family_friendly, replace="family-friendly")}{no_none(mr.customer_rating, postfix=" rated")}{no_none(mr.price_range, postfix=" priced")} {no_none(mr.food)}{no_none(mr.eat_type)}{no_none(mr.near, prefix=" near ")}{no_none(mr.area, prefix="in ")}.'


In [82]:
print('\t', generate(random.choice(meaning_representations)))

	 The Eagle is a family-friendly 3 out of 5 rated moderate priced  English coffee shop  near Burger King in riverside .


#5

1. It would not be hard to add some various phrases, it would only take a bit of time.
2. Would be difficult to deal with the various conjugations of the words.
3. This is a very simple rule based system.
4. Well given the restricted scope it was as difficult as was expected, a better and more complex system would require a lot more thinking
5. If we only want to extract the information with no significant 'new' descriptions classical methods will work fine enough. Neural network will generate more 'novel' descriptions and likely more natural sounding examples. 
6. The rouge score got to about same level as the worst in the table, but BLEU score is not even close.
7. Seq2seq models seem to be quite good in this task, which is to be expected.

#7

1. Was this an interesting description, does it contain all the interesting information.
2. Send person to the restaurant and see if the description was correct.
3. They topped both categories.
4. Well if we are trying to get people to go to restaurants, then naturalness, if we want to give honest evaluations, then optimize for quality.

#6

When creating the improved version, I added restaurant before eat_type field in all things. This increases BLEU score by about 0.007, since likely with the restaurant we get some good n-grams. Since the BLEU score only checks the existence of n-grams, it has no real "understanding" for whether sentences make sense, as a human would evaluate sentences, since the phrase 'restaurant coffee shop' is not really very good but for BLEU it is good since it likely scores some n-grams and gets points.

TypeError: bleu_single() missing 1 required positional argument: 'reference'